Build GCC on CI with GCC, not Clang

Kobzol opened the pull request:
It seems that GCC built with Clang misbehaves. I have tested that cg_gcc tests [pass](https://github.com/rust-lang/rust/actions/runs/13842365913/job/38732750617?pr=138451) on CI with a downloaded GCC that was built in this way.

Prerequisite for https://github.com/rust-lang/rust/pull/138395.

r? @ghost

Kobzol committed:
Create libgccjit.so.0 alias also for CI-downloaded GCC

Kobzol committed:
Change GCC build flags

Kobzol committed:
Use GCC for building GCC

Kobzol commented:
@bors try

bors commented:
:hourglass: Trying commit a1e034a43274d63e932902c5f8b4decfff7136b2 with merge 94e2839e4fb889d033894be1226b47344ce1a17f...
<!-- homu: {"type":"TryBuildStarted","head_sha":"a1e034a43274d63e932902c5f8b4decfff7136b2","merge_sha":"94e2839e4fb889d033894be1226b47344ce1a17f"} -->

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13832551460/job/38713603162)
Build commit: 94e2839e4fb889d033894be1226b47344ce1a17f (`94e2839e4fb889d033894be1226b47344ce1a17f`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13832551460/job/38713603162"},"merge_sha":"94e2839e4fb889d033894be1226b47344ce1a17f"} -->

Kobzol committed:
Do not overwrite original `config.toml` in `opt-dist`

So that follow-up CI commands can proceed normally. It will also avoid overwriting `config.toml` when running opt-dist tests locally.

Kobzol commented:
@bors try

bors commented:
:sunny: Try build successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/13838438291/job/38729930079)
Build commit: fb4237bc14be8ce17d45c5221b52c3e792546032 (`fb4237bc14be8ce17d45c5221b52c3e792546032`)
<!-- homu: {"type":"TryBuildCompleted","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/13838438291/job/38729930079"},"merge_sha":"fb4237bc14be8ce17d45c5221b52c3e792546032"} -->

Kobzol committed:
Store libgccjit.so in a lib directory in the GCC CI tarball

Kobzol commented:
r? @GuillaumeGomez

antoyo commented on src/bootstrap/src/core/build_steps/gcc.rs:
```diff
@@ -210,7 +217,7 @@ fn build_gcc(metadata: &Meta, builder: &Builder<'_>, target: TargetSelection) {
         .env("CXXFLAGS", "-Wno-everything -g -O2")
         .env("CFLAGS", "-Wno-everything -g -O2")
```
Is the `-Wno-everything` still necessary?

Kobzol replied:
I don't think so. Actually, I will add a check that someone doesn't try to compile GCC with Clang, similar to https://github.com/rust-lang/rust/pull/137457.

GuillaumeGomez reviewed (approved):
Apart from @antoyo's comment, looks good to me. Thanks!

Kobzol committed:
Ensure that GCC is not built using Clang, as it misbehaves

Kobzol commented:
Added a commit that checks if we don't compile GCC with Clang, and removed the flags.

GuillaumeGomez reviewed (approved):
Looks all good to me, thanks! r=me once CI pass

Kobzol commented:
@bors r=GuillaumeGomez

bors commented:
:pushpin: Commit bf095f6ecee956b3942625d9480e3a2f13666a67 has been approved by `GuillaumeGomez`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=GuillaumeGomez bf095f6ecee956b3942625d9480e3a2f13666a67 -->
<!-- homu: {"type":"Approved","sha":"bf095f6ecee956b3942625d9480e3a2f13666a67","approver":"GuillaumeGomez","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

nikic commented:
> It seems that GCC built with Clang misbehaves.

Is there a bug report for that?

Kobzol commented:
We think that the specific instance that we hit was https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117952.

But more generally, I expect that building GCC with Clang would require various workarounds and have quirks, we already had to modify compiler flags to even get it to compile.

So just using GCC as the host compiler should be a more robust choice in the future.

nikic commented:
> We think that the specific instance that we hit was https://gcc.gnu.org/bugzilla/show_bug.cgi?id=117952, or something similar.

Thanks, so this is just working around a GCC bug. Just wanted to make sure it's not Clang miscompiling GCC.
